src/Cli/dotnet/commands/dotnet-test/MSBuildConnectionHandler.cs

```diff
@@ -0,0 +1,130 @@
+
+        public MSBuildConnectionHandler(string[] args, TestApplicationActionQueue actionQueue)
+        {
+            _args = args;
+            _actionQueue = actionQueue;
+        }
+
+        public async Task WaitConnectionAsync(CancellationToken token)
+        {
+            VSTestTrace.SafeWriteTrace(() => $"Wait for connection(s) on pipe = {_pipeNameDescription.Name}");
+
+            try
+            {
+                while (true)
+                {
+                    NamedPipeServer pipeConnection = new(_pipeNameDescription, OnRequest, NamedPipeServerStream.MaxAllowedServerInstances, token, skipUnknownMessages: true);
+                    pipeConnection.RegisterAllSerializers();
+
+                    await pipeConnection.WaitConnectionAsync(token);
+
+                    _namedPipeConnections.Add(pipeConnection);
+                }
+            }
+            catch (OperationCanceledException ex) when (ex.CancellationToken == token)
+            {
+                // We are exiting
+            }
+            catch (Exception ex)
+            {
+                if (VSTestTrace.TraceEnabled)
+                {
+                    VSTestTrace.SafeWriteTrace(() => ex.ToString());
+                }
+
+                Environment.FailFast(ex.ToString());
+            }
+        }
+
+        private Task<IResponse> OnRequest(IRequest request)
+        {
+            try
+            {
+                switch (request)
+                {
+                    case Module module:
+                        string modulePath = module.DLLPath;
+                        var testApp = new TestApplication(modulePath, _args);
+                        _actionQueue.Enqueue(testApp);
+                        // Write the test application to the channel
+                        testApp.OnCreated();
+                        break;
+
+                    default:
+                        // If it doesn't match any of the above, throw an exception
+                        throw new NotSupportedException($"Request '{request.GetType()}' is unsupported.");
+                }
+            }
+            catch (Exception ex)
+            {
+                if (VSTestTrace.TraceEnabled)
+                {
+                    VSTestTrace.SafeWriteTrace(() => ex.ToString());
+                }
+
+                Environment.FailFast(ex.ToString());
+            }
+
+            return Task.FromResult((IResponse)VoidResponse.CachedInstance);
+        }
+
+        public int RunWithMSBuild(ParseResult parseResult)
+        {
+            bool containsNoBuild = parseResult.UnmatchedTokens.Any(token => token == CliConstants.NoBuildOptionKey);
+            List<string> msbuildCommandLineArgs =
+            [
+                    $"-t:{(containsNoBuild ? string.Empty : "Build;")}_GetTestsProject",
+                    $"-p:GetTestsProjectPipeName={_pipeNameDescription.Name}",
+                    "-verbosity:q"
+                ];
```

Comment:
Nit: fix indentation either as

```suggestion
            List<string> msbuildCommandLineArgs =
                [
                    $"-t:{(containsNoBuild ? string.Empty : "Build;")}_GetTestsProject",
                    $"-p:GetTestsProjectPipeName={_pipeNameDescription.Name}",
                    "-verbosity:q"
                ];
```

or as

```suggestion
            List<string> msbuildCommandLineArgs =
            [
                $"-t:{(containsNoBuild ? string.Empty : "Build;")}_GetTestsProject",
                $"-p:GetTestsProjectPipeName={_pipeNameDescription.Name}",
                "-verbosity:q"
            ];
```